src/SFML/Window/Win32/WindowImplWin32.cpp

```diff
@@ -747,6 +761,90 @@ void WindowImplWin32::processEvent(UINT message, WPARAM wParam, LPARAM lParam)
 
+        // tell OS whether we're happy with the proposed changes to the window size
+        case WM_SIZING:
+        {
+            RECT* ncsRect = reinterpret_cast<RECT*>(lParam);
+            assert(ncsRect->left < ncsRect->right);
+            assert(ncsRect->top < ncsRect->bottom);
+            auto width  = static_cast<unsigned int>(ncsRect->right - ncsRect->left);
+            auto height = static_cast<unsigned int>(ncsRect->bottom - ncsRect->top);
+            bool isTop  = false; // are we being resized on the top? If not, it's the bottom
+            bool isLeft = false; // are we being resized on the left? If not, it's the right
+            switch (wParam)
+            {
+                case WMSZ_BOTTOMLEFT:
+                    isLeft = true;
+                    break;
+                case WMSZ_TOPLEFT:
+                    isLeft = true;
+                    isTop  = true;
+                    break;
+                case WMSZ_LEFT:
+                    isLeft = true;
+                    break;
+                case WMSZ_TOPRIGHT:
+                    isTop = true;
+                    break;
+                case WMSZ_TOP:
+                    isTop = true;
+                    break;
+            }
+
+
+            if (width < m_minimumSize.x)
+            {
+                // we don't like this
+                if (isLeft)
+                {
+                    ncsRect->left = ncsRect->right - static_cast<LONG>(m_minimumSize.x);
+                }
+                else
+                {
+                    ncsRect->right = ncsRect->left + static_cast<LONG>(m_minimumSize.x);
+                }
+            }
+            else if (width > m_maximumSize.x && m_maximumSize.x != 0)
+            {
+                // we also don't like this either
+                if (isLeft)
+                {
+                    ncsRect->left = ncsRect->right - static_cast<LONG>(m_maximumSize.x);
+                }
+                else
+                {
+                    ncsRect->right = ncsRect->left + static_cast<LONG>(m_maximumSize.x);
+                }
+            }
+
+            if (height < m_minimumSize.y)
+            {
+                // we don't like this
+                if (isTop)
+                {
+                    ncsRect->top = ncsRect->bottom - static_cast<LONG>(m_minimumSize.y);
+                }
+                else
+                {
+                    ncsRect->bottom = ncsRect->top + static_cast<LONG>(m_minimumSize.y);
+                }
+            }
+            else if (height > m_maximumSize.y && m_maximumSize.y != 0)
+            {
+                // we also don't like this
+                if (isTop)
+                {
+                    ncsRect->top = ncsRect->bottom - static_cast<LONG>(m_maximumSize.y);
+                }
+                else
+                {
+                    ncsRect->bottom = ncsRect->top + static_cast<LONG>(m_maximumSize.y);
+                }
+            }
```

Comment:
This is very verbose. Please reuse some code; the branches of `if` and `else if` are almost identical.

Also, could you make use of [`std::clamp()`](https://en.cppreference.com/w/cpp/algorithm/clamp), or at least `std::min/max()`?

Last, what is the "we don't like this" comment intended to mean? 🙂

Reply:
"We don't like this" => The resize size proposed to us from the operating system violates the minimum or maximum constraints that the consuming developer has specified, so we must take action to adjust the rectangle to comply with the minimum or maximum values.

Reply:
`std::min(a, b)` and `std::max(a, b)` just return the smaller or larger value. They just got added in C++11/14.
`std::clamp(value, low, high)` was added in C++17 and returns `value`, but ensuring it's neither lower nor higher than `low` and `high`.

Macro equivalents:
```cpp
#define min(a, b) ((a) < (b) ? (a) : (b))
#define max(a, b) ((a) > (b) ? (a) : (b))
#define clamp(v, l, h) ((v) < (l) ? (l) : ((v) > (h) ? (h) : (v)))
```
For example, to set the new bottom edge:
```cpp
ncsRect->bottom = ncsRect->top + std::clamp(height, m_minimumSize.y, m_maximumSize.y);
```
Or to set the new top edge:
```cpp
ncsRect->top = ncsRect->bottom - std::clamp(height, m_minimumSize.y, m_maximumSize.y);
```

Reply:
This is still unadressed. I would not solve it with macros though.

Reply:
I cleaned up this implementation a bit but I still can't figure out `std::min`/`std::max`/`std::clamp` would help. The problem is that we have to do different things based on the direction from which resizing is occurring. We're not simply taking a vector and clamping it a min and max. We have to take into account the side of the window in which resizing is occurring.

Regardless, shouldn't we be using something like this: https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-getminmaxinfo?redirectedfrom=MSDN

Surely Windows provides ways of telling it what the minimum and maximum sizes are instead of implementing all of this logic ourselves, right?